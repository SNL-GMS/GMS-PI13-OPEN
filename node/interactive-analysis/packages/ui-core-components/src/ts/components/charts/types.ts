import { SohTypes } from '@gms/common-graphql';
import { Point } from '@gms/ui-util';

// tslint:disable: no-magic-numbers
export const EVERY_20_SECONDS_FOR_A_MINUTE = 3;
export const EVERY_20_SECONDS_FOR_AN_HOUR = EVERY_20_SECONDS_FOR_A_MINUTE * 60;
export const EVERY_20_SECONDS_FOR_A_DAY = EVERY_20_SECONDS_FOR_AN_HOUR * 24;
export const EVERY_20_SECONDS_FOR_A_WEEK = EVERY_20_SECONDS_FOR_A_DAY * 7;
export const EVERY_20_SECONDS_FOR_A_MONTH = EVERY_20_SECONDS_FOR_A_DAY * 30;
// tslint:enable: no-magic-numbers

/** defines a line */
export type Line = Point[];

export interface LineDefinition {
  id: string | number;
  color: string;
  value: Line;
}

export interface BarDefinition {
  id: string | number;
  color: string;
  value: BarValue;
}

export interface BarValue extends Point {
  quietUntilMs?: number;
  quietDurationMs?: number;
  channelStatus?: SohTypes.SohStatusSummary;
}

/** defines a domain tuple */
export type DomainTuple = [number, number];

/** defines a domain */
export interface Domain {
  x: DomainTuple;
  y: DomainTuple;
}

/** defines the chart padding */
export interface ChartPadding {
  top: number;
  left: number;
  bottom: number;
  right: number;
}

/** The base chart props */
export interface BaseChartProps {
  id?: string | number;
  classNames?: string;
  widthPx: number;
  heightPx: number;
  minHeightPx?: number;
  padding?: ChartPadding;
  domain?: Domain;
}

/** The axis props */
export interface AxisProps {
  rotateAxis?: boolean;
  suppressYAxis?: boolean;
  suppressXAxis?: boolean;
  yAxisLabel?: string;
  xAxisLabel?: string;
  yTickCount?: number;
  xTickCount?: number;
  xTickValues?: string[] | number[];
  yTickValues?: string[] | number[];
  xTickTooltips?: string[];
  yTickTooltips?: string[];
  barDefs?: BarDefinition[];
  disabled?: Disabled;
  onContextMenuBar?(e: any, datum: any): void;
  onContextMenuBarLabel?(e: any, index: number): void;
  yTickFormat?(value: string | number): string | number;
  xTickFormat?(value: string | number): string | number;
}

/** The line chart props */
export interface LineChartProps extends BaseChartProps, AxisProps {
  lineDefs: LineDefinition[];
  interpolation?: any;
  stepChart?: boolean;
  suppressScrolling?: boolean;
  zoomDomain?: Domain;
  domain?: Domain;
  onZoomDomainChange?(domain: Domain): void;
}

/** The bar chart props */
export interface BarChartProps extends BaseChartProps, AxisProps {
  maxBarWidth: number;
  minBarWidth: number;
  categories: { x: string[]; y: string[] };
  barDefs: BarDefinition[];
  scrollBrushColor?: string;
  thresholdsMarginal?: number[];
  thresholdsBad?: number[];
  disabled?: Disabled;
  dataComponent?: React.Component;
}

interface Disabled {
  xTicks: {
    disabledColor: string;
    disabledCondition(tick: { value: number; tooltip: string }): boolean;
  };
}
