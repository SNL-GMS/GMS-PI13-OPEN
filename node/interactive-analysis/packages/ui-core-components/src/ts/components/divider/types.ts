/**
 * Divider State
 */
// tslint:disable-next-line: no-empty-interface
export interface DividerState {}

/**
 * Divider Props
 */
// tslint:disable-next-line: no-empty-interface
export interface DividerProps {}
