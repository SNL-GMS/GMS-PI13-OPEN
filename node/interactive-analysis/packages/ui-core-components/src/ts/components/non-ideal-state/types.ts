/**
 * A NonIdealStateDefinition for a given type T pairs the non-ideal state component
 * with a condition that determines whether to render it.
 */
export interface NonIdealStateDefinition<T> {
  // The JSX to render for the non ideal state
  element: JSX.Element;
  // take in the component's props and return true if the non ideal state should be rendered.
  condition(props: T): boolean;
}
