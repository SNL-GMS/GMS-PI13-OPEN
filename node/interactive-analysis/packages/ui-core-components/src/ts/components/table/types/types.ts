import { CellRendererParams, ColumnDefinition, ValueGetterParams } from 'src/ts/ui-core-components';

/**
 * ! Here is where types are defined so that when using the table component have access to more specific types
 * ! that are generic
 */

/**
 * Column definition - value is of type string
 */
export type StringColumnDefinition = ColumnDefinition<{ id: string }, {}, string, {}, {}>;

/**
 * Column definition - value is of type number
 */
export type NumberColumnDefinition = ColumnDefinition<{ id: string }, {}, number, {}, {}>;

/**
 * Column value getter params - value is of type string
 */
export type StringValueGetterParams = ValueGetterParams<{ id: string }, {}, string, {}, {}>;

/**
 * Column value getter - value is of type string
 */
export type StringValueGetter = (params: StringValueGetterParams) => string;

/**
 * Column value getter params - value is of type number
 */
export type NumberValueGetterParams = ValueGetterParams<{ id: string }, {}, number, {}, {}>;
/**
 * Column value getter - value is of type number
 */
export type NumberValueGetter = (params: NumberValueGetterParams) => number;

/**
 * Cell renderer params - value is of type string
 */
export type StringCellRendererParams = CellRendererParams<{ id: string }, {}, string, {}, {}>;

/**
 * Cell renderer params - value is of type number
 */
export type NumberCellRendererParams = CellRendererParams<{ id: string }, {}, number, {}, {}>;
