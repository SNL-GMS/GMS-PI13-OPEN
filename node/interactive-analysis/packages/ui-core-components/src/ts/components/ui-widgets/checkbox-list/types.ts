import Immutable from 'immutable';

export interface CheckboxListProps {
  checkboxEnum: any;
  enumToCheckedMap: Map<any, boolean>;
  enumKeysToDisplayStrings?: Map<string, string>;
  enumToColorMap?: Map<any, string>;
  onChange(map: Map<any, boolean>);
}

export interface SimpleCheckboxListProps {
  checkBoxListEntries: CheckboxListEntry[];
  onChange?(name: string): void;
}

export interface SimpleCheckboxListState {
  checkboxEntriesMap: Immutable.Map<string, CheckboxListEntry>;
}

export interface CheckboxListEntry {
  name: string;
  color?: string;
  isChecked: boolean;
}
