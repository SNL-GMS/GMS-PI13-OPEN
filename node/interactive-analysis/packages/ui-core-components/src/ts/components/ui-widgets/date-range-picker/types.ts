import { DateRange, TimePrecision } from '@blueprintjs/datetime';

/** Defines a trend */
export interface Trend {
  /** the description of the trend ... i.e. `last 24 hours` */
  description: string;
  /** the number of milliseconds */
  value: number;
}

// tslint:disable-next-line:no-empty-interface
export interface DateRangePickerProps {
  startDate: Date;
  endDate: Date;
  defaultTrends: Trend[];
  renderStacked?: boolean;
  onNewInterval(startDate: Date, endDate: Date);
  onApply?(startDate: Date, endDate: Date);
}
export interface DateRangePickerState {
  allowSingleDayRange?: boolean;
  singleMonthOnly?: boolean;
  contiguousCalendarMonths?: boolean;
  dateRange?: DateRange;
  maxDateIndex?: number;
  minDateIndex?: number;
  reverseMonthAndYearMenus?: boolean;
  shortcuts?: boolean;
  timePrecision?: TimePrecision;
  selectedDates: DateRange;
  showDatePicker: boolean;
}
