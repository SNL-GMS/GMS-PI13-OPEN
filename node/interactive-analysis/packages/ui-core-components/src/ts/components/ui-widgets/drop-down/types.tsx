export interface DropDownProps {
  value: string;
  dropDownItems: any;
  dropdownText?: any;
  widthPx?: number;
  disabled?: boolean;
  title?: string;
  custom?: boolean;
  'data-cy'?: string;
  className?: string;
  onMaybeValue(value: any);
}
