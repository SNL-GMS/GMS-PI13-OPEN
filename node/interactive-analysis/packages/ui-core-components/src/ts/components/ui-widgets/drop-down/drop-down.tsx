import { HTMLSelect } from '@blueprintjs/core';
import React from 'react';
import { WidgetTypes } from '..';
import { DropDownProps } from './types';

/**
 * Drop Down menu
 */
const UNSELECTABLE_CUSTOM_VALUE = 'UNSELECTED_CUSTOM_VALUE';

export class DropDown extends React.Component<DropDownProps, WidgetTypes.WidgetState> {
  private constructor(props) {
    super(props);
    this.state = {
      value: this.props.value,
      isValid: true
    };
  }
  /**
   * React component lifecycle.
   */
  public render() {
    const minWidth = `${this.props.widthPx}px`;
    const altStyle = {
      minWidth,
      width: minWidth
    };
    return (
      <HTMLSelect
        title={`${this.props.title}`}
        disabled={this.props.disabled}
        style={this.props.widthPx !== undefined ? altStyle : undefined}
        className={this.props.className}
        onChange={e => {
          const input = e.target.value;
          if (this.props.custom && input === UNSELECTABLE_CUSTOM_VALUE) {
            return;
          }
          this.props.onMaybeValue(input);
        }}
        data-cy={this.props['data-cy']}
        value={this.props.custom ? UNSELECTABLE_CUSTOM_VALUE : this.props.value}
      >
        {this.createDropdownItems(this.props.dropDownItems, this.props.dropdownText)}
        {this.props.custom ? (
          <option key={UNSELECTABLE_CUSTOM_VALUE} value={UNSELECTABLE_CUSTOM_VALUE}>
            Custom
          </option>
        ) : null}
      </HTMLSelect>
    );
  }

  /**
   * Creates the HTML for the dropdown items for the type input
   *
   */
  private readonly createDropdownItems = (enumOfOptions: any, dropdownText: any): JSX.Element[] => {
    const items: any[] = [];
    Object.keys(enumOfOptions).forEach(type => {
      items.push(
        <option key={type} value={enumOfOptions[type]}>
          {dropdownText ? dropdownText[type] : enumOfOptions[type]}
        </option>
      );
    });
    return items;
  }
}
