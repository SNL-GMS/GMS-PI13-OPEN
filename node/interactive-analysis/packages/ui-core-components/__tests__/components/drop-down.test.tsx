import React from 'react';
import { DropDown } from '../../src/ts/components/ui-widgets/drop-down';
import { DropDownProps } from '../../src/ts/components/ui-widgets/drop-down/types';

// tslint:disable-next-line:no-implicit-dependencies no-var-requires no-require-imports
const Enzyme = require('enzyme');

// set up window alert and open so we don't see errors
(window as any).alert = jest.fn();
(window as any).open = jest.fn();

function flushPromises(): any {
  return new Promise(setImmediate);
}

describe('Core drop down', () => {
  enum TEST_ENUM {
    test = 'test',
    foo = 'foo',
    bar = 'bar'
  }
  let testValue = TEST_ENUM.bar;
  const props: DropDownProps = {
    dropDownItems: TEST_ENUM,
    value: testValue,
    widthPx: 120,
    disabled: false,
    onMaybeValue: value => {
      testValue = value;
    }
  };
  const mockDropdown = Enzyme.shallow(<DropDown {...props} />);

  it('Renders', () => {
    expect(mockDropdown).toMatchSnapshot();
  });
  it('Can have a value set', () => {
    mockDropdown.find('HTMLSelect').prop('onChange')({
      target: { value: TEST_ENUM.foo }
    });
    flushPromises();
    expect(testValue).toEqual(TEST_ENUM.foo);
  });
  it('Can render custom value', () => {
    const props2: DropDownProps = {
      dropDownItems: TEST_ENUM,
      value: testValue,
      widthPx: 120,
      disabled: false,
      onMaybeValue: value => {
        testValue = value;
      },
      custom: true
    };
    const mockDropdownCustom = Enzyme.shallow(<DropDown {...props2} />);
    expect(mockDropdownCustom).toMatchSnapshot();
  });
});
