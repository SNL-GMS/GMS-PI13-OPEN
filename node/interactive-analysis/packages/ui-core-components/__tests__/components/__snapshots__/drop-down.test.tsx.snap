// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Core drop down Can render custom value 1`] = `
<HTMLSelect
  disabled={false}
  onChange={[Function]}
  style={
    Object {
      "minWidth": "120px",
      "width": "120px",
    }
  }
  title="undefined"
  value="UNSELECTED_CUSTOM_VALUE"
>
  <option
    key="test"
    value="test"
  >
    test
  </option>
  <option
    key="foo"
    value="foo"
  >
    foo
  </option>
  <option
    key="bar"
    value="bar"
  >
    bar
  </option>
  <option
    key="UNSELECTED_CUSTOM_VALUE"
    value="UNSELECTED_CUSTOM_VALUE"
  >
    Custom
  </option>
</HTMLSelect>
`;

exports[`Core drop down Renders 1`] = `
<HTMLSelect
  disabled={false}
  onChange={[Function]}
  style={
    Object {
      "minWidth": "120px",
      "width": "120px",
    }
  }
  title="undefined"
  value="bar"
>
  <option
    key="test"
    value="test"
  >
    test
  </option>
  <option
    key="foo"
    value="foo"
  >
    foo
  </option>
  <option
    key="bar"
    value="bar"
  >
    bar
  </option>
</HTMLSelect>
`;
